Add spacy-llm docs to webiste (#12782)
explosion/spaCy: website/docs/usage/large-language-models.mdx

  functionality to parse the resulting response back into structured fields on
  spaCy's [Doc](https://spacy.io/api/doc) objects.
- A [**model**](#models) defining the model to use and how to connect to it.
  Note that `spacy-llm` supports both access to external APIs (such as OpenAI)
  as well as access to self-hosted open-source LLMs (such as using Dolly through
  Hugging Face).

Moreover, `spacy-llm` exposes a customizable [**caching**](#cache) functionality
to avoid running the same document through an LLM service (be it local or
through a REST API) more than once.

Finally, you can choose to save a stringified version of LLM prompts/responses
within the `Doc.user_data["llm_io"]` attribute by setting `save_io` to `True`.
`Doc.user_data["llm_io"]` is a dictionary containing one entry for every LLM
component within the spaCy pipeline. Each entry is itself a dictionary, with two
keys: `prompt` and `response`.

A note on `validate_types`: by default, `spacy-llm` checks whether the
signatures of the `model` and `task` callables are consistent with each other
and emits a warning if they don't. `validate_types` can be set to `False` if you

> want to disable this behavior.

Reviewer: It feels like quite a bit of text got duplicated between the usage and the api pages, I would reduce the duplication as much as we can.

in [`ty.py`](https://github.com/spacy-llm/spacy_llm/ty.py). It needs to define a
`generate_prompts` function and a `parse_responses` function.

Moreover, the task may define an optional
[`scorer` method](https://spacy.io/api/scorer#score). It should accept an
iterable of `Example`s as input and return a score dictionary. If the `scorer`
method is defined, `spacy-llm` will call it to evaluate the component.

| Component                                                                   | Description                                                                                                                                                           |
| --------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| [`task.generate_prompts`](/api/large-language-models#task-generate-prompts) | Takes a collection of documents, and returns a collection of "prompts", which can be of type `Any`.                                                                   |

> | [`task.parse_responses`](/api/large-language-models#task-parse-responses)   | Takes a collection of LLM responses and the original documents, parses the responses into structured information, and sets the annotations on the documents.          |

Reviewer: The table would look nicer if we can have these first two methods somewhere else, maybe inbetween the previous 2 paragraphs?

| [Normalizer functions](/api/large-language-models#normalizer-functions) | These functions provide simple normalizations for string comparisons, e.g. between a list of specified labels and a label given in the raw text of the LLM response.                                                                                                                 |

## Ongoing work {id="ongoing-work"}

In the near future, we will

- Add more example tasks
- Support a broader range of models
- Provide more example use-cases and tutorials
- Make the built-in tasks easier to customize via Jinja templates to define the
  instructions & examples

PRs are always welcome!

## Reporting issues {id="issues"}

If you have questions regarding the usage of `spacy-llm`, or want to give us
feedback after giving it a spin, please use the
[discussion board](https://github.com/explosion/spaCy/discussions). Bug reports
can be filed on the

> [spaCy issue tracker](https://github.com/explosion/spaCy/issues). Thank you!

Reviewer: This looks to me more like "readme" sections that things we should have in the spacy.io docs

Reviewer 2: Yes, we can ditch this. This will remain in the readme.

---
title: Large Language Models
teaser: Integrating LLMs into structured NLP pipelines
menu:
  - ['Motivation', 'motivation']
  - ['Install', 'install']
  - ['Usage', 'usage']
  - ['Logging', 'logging']
  - ['API', 'api']
  - ['Tasks', 'tasks']

>   - ['Backends', 'backends']

Reviewer 2: ```suggestion
  - ['Models', 'models']
```

You are an expert Text Classification system. Your task is to accept Text as input
and provide a category for the text based on the predefined labels.

Classify the text below to any of the following labels: COMPLIMENT, INSULT
The task is non-exclusive, so you can provide more than one label as long as
they're comma-delimited. For example: Label1, Label2, Label3.
Do not put any other text in your answer, only one or more of the provided labels with nothing before or after.
If the text cannot be classified into any of the provided labels, answer `==NONE==`.

Here is the text that needs classification

Text:
'''
You look gorgeous!
'''

Model response for doc: You look gorgeous!
COMPLIMENT
```

`print(doc.cats)` to standard output should look like:

```
{'COMPLIMENT': 1.0, 'INSULT': 0.0}
```

## API {id="api"}

`spacy-llm` exposes a `llm` factory with
[configurable settings](api/large-language-models#config).

An `llm` component is defined by two main settings:

- A [**task**](#tasks), defining the prompt to send to the LLM as well as the
  functionality to parse the resulting response back into structured fields on
  spaCy's [Doc](https://spacy.io/api/doc) objects.

> - A [**backend**](#backends) defining the model to use and how to connect to it.

Reviewer 2: ```suggestion
- A [**model**](#models) defining the model to use and how to connect to it.
```

  Note that `spacy-llm` supports both access to external APIs (such as OpenAI)
  as well as access to self-hosted open-source LLMs (such as using Dolly through
  Hugging Face).

Moreover, `spacy-llm` exposes a customizable [**caching**](#cache) functionality
to avoid running the same document through an LLM service (be it local or
through a REST API) more than once.

Finally, you can choose to save a stringified version of LLM prompts/responses
within the `Doc.user_data["llm_io"]` attribute by setting `save_io` to `True`.
`Doc.user_data["llm_io"]` is a dictionary containing one entry for every LLM
component within the spaCy pipeline. Each entry is itself a dictionary, with two
keys: `prompt` and `response`.

A note on `validate_types`: by default, `spacy-llm` checks whether the

> signatures of the `backend` and `task` callables are consistent with each other

Reviewer 2: ```suggestion
signatures of the `model` and `task` callables are consistent with each other
```

| [`spacy.NoOp.v1`](/api/large-language-models#noop-v1)                       | This task is only useful for testing - it tells the LLM to do nothing, and does not set any fields on the `docs`.                                                     |

### Model {id="models"}

A _model_ defines which LLM model to query, and how to query it. It can be a
simple function taking a collection of prompts (consistent with the output type
of `task.generate_prompts()`) and returning a collection of responses
(consistent with the expected input of `parse_responses`). Generally speaking,
it's a function of type `Callable[[Iterable[Any]], Iterable[Any]]`, but specific
implementations can have other signatures, like
`Callable[[Iterable[str]], Iterable[str]]`.

All built-in models are registered in `llm_models`. If no model is specified,
the repo currently connects to the `OpenAI` API by default using REST, and
accesses the `"gpt-3.5-turbo"` model.

Currently three different approaches to use LLMs are supported:

> 1. `spacy-llm`s native REST backend. This is the default for all hosted models

Reviewer 2: ```suggestion
1. `spacy-llm`s native REST interface. This is the default for all hosted models
```

   (e. g. OpenAI, Cohere, Anthropic, ...).
2. A HuggingFace integration that allows to run a limited set of HF models
   locally.
3. A LangChain integration that allows to run any model supported by LangChain
   (hosted or locally).

Approaches 1. and 2 are the default for hosted model and local models,
respectively. Alternatively you can use LangChain to access hosted or local
models by specifying one of the models registered with the `langchain.` prefix.

<Infobox>

> _Why LangChain if there are also are a native REST and a HuggingFace backend? When should I use what?_

Reviewer 2: ```suggestion
_Why LangChain if there are also are a native REST and a HuggingFace interface? When should I use what?_
```

---
title: Large Language Models
teaser: Integrating LLMs into structured NLP pipelines
menu:
  - ['Motivation', 'motivation']
  - ['Install', 'install']
  - ['Usage', 'usage']
  - ['Logging', 'logging']
  - ['API', 'api']
  - ['Tasks', 'tasks']
  - ['Models', 'models']
---

[The spacy-llm package](https://github.com/explosion/spacy-llm) integrates Large

> Language Models (LLMs) into [spaCy](https://spacy.io), featuring a modular

Reviewer: ```suggestion
Language Models (LLMs) into spaCy pipelines, featuring a modular
```

system for **fast prototyping** and **prompting**, and turning unstructured
responses into **robust outputs** for various NLP tasks, **no training data**
required.

- Serializable `llm` **component** to integrate prompts into your pipeline
- **Modular functions** to define the [**task**](#tasks) (prompting and parsing)
  and [**model**](#models) (model to use)
- Support for **hosted APIs** and self-hosted **open-source models**
- Integration with [`LangChain`](https://github.com/hwchase17/langchain)
- Access to
  **[OpenAI API](https://platform.openai.com/docs/api-reference/introduction)**,
  including GPT-4 and various GPT-3 models

> - Built-in support for **open-source
>   [Dolly](https://huggingface.co/databricks)** models hosted on Hugging Face

Reviewer: ```suggestion
- Built-in support for various **open-source** models hosted on [Hugging Face](https://huggingface.co/)
```

> - Usage examples for **Named Entity Recognition** and **Text Classification**

Reviewer: ```suggestion
- Usage examples for standard NLP tasks such as **Named Entity Recognition** and **Text Classification**
```

- Easy implementation of **your own functions** via

>   [spaCy's registry](https://spacy.io/api/top-level#registry) for custom

Reviewer: ```suggestion
  the [registry](https://spacy.io/api/top-level#registry) for custom
```

Reviewer: PS Victoria, here's one of those absolute links ;-)

  prompting, parsing and model integrations

## Motivation {id="motivation"}

Large Language Models (LLMs) feature powerful natural language understanding
capabilities. With only a few (and sometimes no) examples, an LLM can be
prompted to perform custom NLP tasks such as text categorization, named entity
recognition, coreference resolution, information extraction and more.

[spaCy](https://spacy.io) is a well-established library for building systems
that need to work with language in various ways. spaCy's built-in components are

> generally powered by supervised learning or rule-based approaches.

Reviewer: It feels weird to keep this paragraph in here?

is fully justified. If you want a system that can synthesize information from
multiple documents in subtle ways and generate a nuanced summary for you, bigger
is better. However, even if your production system needs an LLM for some of the
task, that doesn't mean you need an LLM for all of it. Maybe you want to use a
cheap text classification model to help you find the texts to summarize, or
maybe you want to add a rule-based system to sanity check the output of the
summary. These before-and-after tasks are much easier with a mature and
well-thought-out library, which is exactly what spaCy provides.

## Install {id="install"}

`spacy-llm` will be installed automatically in future spaCy versions. For now,
you can run the following in the same virtual environment where you already have
`spacy` [installed](https://spacy.io/usage).

> ⚠️ This package is still experimental and it is possible that changes made to
> the interface will be breaking in minor version updates.

```bash
python -m pip install spacy-llm
```

## Usage {id="usage"}

The task and the model have to be supplied to the `llm` pipeline component using

> [spaCy's config system](https://spacy.io/api/data-formats#config). This package

Reviewer: ```suggestion
the [config system](https://spacy.io/api/data-formats#config). This package
```

[DollyHF](/api/large-language-models#dolly) section.

Create a config file `config.cfg` containing at least the following (or see the
full example
[here](https://github.com/explosion/spacy-llm/tree/main/usage_examples/ner_dolly)):

```ini
[nlp]
lang = "en"
pipeline = ["llm"]

[components]

[components.llm]
factory = "llm"

[components.llm.task]
@llm_tasks = "spacy.NER.v2"
labels = ["PERSON", "ORGANISATION", "LOCATION"]

[components.llm.model]
@llm_models = "spacy.Dolly.v1"
# For better performance, use dolly-v2-12b instead
name = "dolly-v2-3b"
```

Now run:

```python
from spacy_llm.util import assemble

nlp = assemble("config.cfg")
doc = nlp("Jack and Jill rode up the hill in Les Deux Alpes")
print([(ent.text, ent.label_) for ent in doc.ents])
```

Note that Hugging Face will download the `"databricks/dolly-v2-3b"` model the
first time you use it. You can
[define the cached directory](https://huggingface.co/docs/huggingface_hub/main/en/guides/manage-cache)
by setting the environmental variable `HF_HOME`. Also, you can upgrade the model
to be `"databricks/dolly-v2-12b"` for better performance.

### Example 3: Create the component directly in Python {id="example-3"}

> The `llm` component behaves as any other spaCy component does, so adding it to

Reviewer: ```suggestion
The `llm` component behaves as any other component does, so adding it to
```

an existing pipeline follows the same pattern:

```python
import spacy

nlp = spacy.blank("en")
nlp.add_pipe(
    "llm",
    config={
        "task": {
            "@llm_tasks": "spacy.NER.v2",
            "labels": ["PERSON", "ORGANISATION", "LOCATION"]
        },
        "model": {
            "@llm_models": "spacy.gpt-3.5.v1",
        },
    },
)
nlp.initialize()
doc = nlp("Jack and Jill rode up the hill in Les Deux Alpes")
print([(ent.text, ent.label_) for ent in doc.ents])
```

Note that for efficient usage of resources, typically you would use
[`nlp.pipe(docs)`](https://spacy.io/api/language#pipe) with a batch, instead of
calling `nlp(doc)` with a single document.

### Example 4: Implement your own custom task {id="example-4"}

To write a [`task`](#tasks), you need to implement two functions:

> `generate_prompts` that takes a list of spaCy [`Doc`](https://spacy.io/api/doc)

Reviewer: ```suggestion
`generate_prompts` that takes a list of [`Doc`](https://spacy.io/api/doc)
```

```
Generated prompt for doc: You look gorgeous!

You are an expert Text Classification system. Your task is to accept Text as input
and provide a category for the text based on the predefined labels.

Classify the text below to any of the following labels: COMPLIMENT, INSULT
The task is non-exclusive, so you can provide more than one label as long as
they're comma-delimited. For example: Label1, Label2, Label3.
Do not put any other text in your answer, only one or more of the provided labels with nothing before or after.
If the text cannot be classified into any of the provided labels, answer `==NONE==`.

Here is the text that needs classification

Text:
'''
You look gorgeous!
'''

Model response for doc: You look gorgeous!
COMPLIMENT
```

`print(doc.cats)` to standard output should look like:

```
{'COMPLIMENT': 1.0, 'INSULT': 0.0}
```

## API {id="api"}

`spacy-llm` exposes a `llm` factory with
[configurable settings](/api/large-language-models#config).

An `llm` component is defined by two main settings:

- A [**task**](#tasks), defining the prompt to send to the LLM as well as the
  functionality to parse the resulting response back into structured fields on

>   spaCy's [Doc](https://spacy.io/api/doc) objects.

Reviewer: ```suggestion
  the [Doc](https://spacy.io/api/doc) objects.
```

- A [**model**](#models) defining the model to use and how to connect to it.
  Note that `spacy-llm` supports both access to external APIs (such as OpenAI)
  as well as access to self-hosted open-source LLMs (such as using Dolly through
  Hugging Face).

Moreover, `spacy-llm` exposes a customizable [**caching**](#cache) functionality
to avoid running the same document through an LLM service (be it local or
through a REST API) more than once.

Finally, you can choose to save a stringified version of LLM prompts/responses
within the `Doc.user_data["llm_io"]` attribute by setting `save_io` to `True`.
`Doc.user_data["llm_io"]` is a dictionary containing one entry for every LLM

> component within the spaCy pipeline. Each entry is itself a dictionary, with two

Reviewer: ```suggestion
component within the `nlp` pipeline. Each entry is itself a dictionary, with two
```

keys: `prompt` and `response`.

A note on `validate_types`: by default, `spacy-llm` checks whether the
signatures of the `model` and `task` callables are consistent with each other
and emits a warning if they don't. `validate_types` can be set to `False` if you
want to disable this behavior.

### Tasks {id="tasks"}

A _task_ defines an NLP problem or question, that will be sent to the LLM via a
prompt. Further, the task defines how to parse the LLM's responses back into

> structured information. All tasks are registered in spaCy's `llm_tasks`

Reviewer: ```suggestion
structured information. All tasks are registered in the `llm_tasks`
```

| [`spacy.Lemma.v1`](/api/large-language-models#lemma-v1)                 | The `Lemma.v1` task lemmatizes the provided text and updates the `lemma_` attribute in the doc's tokens accordingly.                                                  |
| [`spacy.Sentiment.v1`](/api/large-language-models#sentiment-v1)         | Performs sentiment analysis on provided texts.                                                                                                                        |
| [`spacy.NoOp.v1`](/api/large-language-models#noop-v1)                   | This task is only useful for testing - it tells the LLM to do nothing, and does not set any fields on the `docs`.                                                     |

#### Providing examples for few-shot prompts {id="few-shot-prompts"}

All built-in tasks support few-shot prompts, i. e. including examples in a
prompt. Examples can be supplied in two ways: (1) as a separate file containing
only examples or (2) by initializing `llm` with a `get_examples()` callback

> (like any other spaCy pipeline component).

Reviewer: ```suggestion
(like any other pipeline component).
```
